### PR TITLE
support non-collection return types

### LINKED ISSUE #1350: Datastore findByX always expects a list returned
Version: 1.1.0.RC2
JDK: 8
Spring Boot Version: 2.1.1.RELEASE
---

I have a repository which looks like;

```
public interface UserRepository extends DatastoreRepository<User, String> {
    Optional<User> findOneByUsername(String username);
}
```

When calling `findOneByUsername` I expect a max of one record to be returned. In actual fact, a list is being returned and I get this error;

```
Unable to convert class java.util.ArrayList to class io.jacent.api.model.User
```

I have also tried `findByUsername` (which the Spring Data docs say should be fine given the return value is singular) with the same results. The expected behaviour is demonstrated when using a JPA repository.

### LINKED ISSUE #1471: Getting a single entry findOneBy... / findFirstBy... results in DatastoreDataException
Hello,

I want to load a single entry by a repository like this:

```
public interface UserRepository extends DatastoreRepository<User, Long> {
	User findFirstByGoogleAccountId(String googleAccountId);
}
```

But a call of this method leads to this exception and stacktrace:
```
org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class java.util.ArrayList to class repositories.user.User                                                   
     at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnReadSingle(TwoStepsConversions.java:231) ~[org_springframework_cloud_spring_cloud_gcp_data_datastore.jar:1.1.0.RE 
     at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnRead(TwoStepsConversions.java:164) ~[org_springframework_cloud_spring_cloud_gcp_data_datastore.jar:1.1.0.RELEASE] 
     at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnRead(TwoStepsConversions.java:105) ~[org_springframework_cloud_spring_cloud_gcp_data_datastore.jar:1.1.0.RELEASE] 
     at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.convertResultCollection(PartTreeDatastoreQuery.java:211) ~[org_springframework_cloud_spring_cloud_gcp_data_datastor 
     at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.execute(PartTreeDatastoreQuery.java:180) ~[org_springframework_cloud_spring_cloud_gcp_data_datastore.jar:1.1.0.RELE 
     at org.springframework.cloud.gcp.data.datastore.repository.query.PartTreeDatastoreQuery.execute(PartTreeDatastoreQuery.java:131) ~[org_springframework_cloud_spring_cloud_gcp_data_datastore.jar:1.1.0.RELE 
     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:605) ~[org_springframework_data_spring_data_commons.jar: 
     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[org_springframework_data_spring_data_commo 
     at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[org_springframework_data_spring_data_commons.jar:2. 
     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[org_springframework_spring_aop.jar:5.1.5.RELEASE]                                            
     at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[org_springframework_data_spring_data_commons.jar:2.1.4.RELEASE]      
     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[org_springframework_spring_aop.jar:5.1.5.RELEASE]                                            
     at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[org_springframework_spring_aop.jar:5.1.5.RELEASE]                                          
     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[org_springframework_spring_aop.jar:5.1.5.RELEASE]                                            
     at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[org_springframework_data_spring_data_ 
     at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[org_springframework_spring_aop.jar:5.1.5.RELEASE]                                            
     at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[org_springframework_spring_aop.jar:5.1.5.RELEASE]                                                             
     at com.sun.proxy.$Proxy59.findFirstByGoogleAccountId(Unknown Source) ~                                                                                                                                      
```

Neither does `findOneBy...` nor result type `Optional<User>` works.
Isn't that the usual way to load a single entry in JPA?

### PR DESCRIPTION
**Pull request by dmitry-s**
fixes #1471 
fixes #1350

### CODE REVIEW

**Comment by dmitry-s** [comment]
I think that it should be expected to work, since it is part of spring data. But maybe a couple of examples could be useful.

File: docs/src/main/asciidoc/datastore.adoc [hunk 1]
```diff
@@ -855,6 +855,15 @@ Filtering parameters can be of types supported by your configured custom convert
...
   List<Trader> findByAction(String action);
 
+  //throws an exception if no results
+  Trader findOneByAction(String action);
+
+  //because of the annotation, returns null if no results
+  @Nullable
+  Trader getByAction(String action);
+
+  Oprional<Trader> getOneByAction(String action);
```
Comment (ChengyuanZhao) [inline docs/src/main/asciidoc/datastore.adoc]: typo in "Oprional"

File: spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQueryTests.java [hunk 2]
```diff
@@ -637,15 +642,99 @@ public void existShouldBeFalseWhenResultSetIsEmpty() throws NoSuchMethodExceptio
...
+
 	private void queryWithMockResult(String queryName, List results, Method m) {
+		queryWithMockResult(queryName, results, m, false);
+	}
+
+	private void queryWithMockResult(String queryName, List results, Method m, boolean mockOprionalNullable) {
 		when(this.queryMethod.getName()).thenReturn(queryName);
 		doReturn(new DefaultParameters(m))
 				.when(this.queryMethod).getParameters();
+		if (mockOprionalNullable) {
```
Comment (ChengyuanZhao) [inline spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQueryTests.java]: typo here "Oprional"

File: docs/src/main/asciidoc/datastore.adoc (hunk 1 above)
Comment (ChengyuanZhao) [inline docs/src/main/asciidoc/datastore.adoc]:
```suggestion
  Optional<Trader> getOneByAction(String action);
```

File: spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQueryTests.java (hunk 2 above)
Comment (ChengyuanZhao) [inline spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQueryTests.java]:
```suggestion
		if (mockOptionalNullable) {
```

**Pull request merged**